File: backend/internal/email/email.go
// Package email ports app/services/email_service.py — sends
// transactional emails (team invites, GST sync summaries) via SMTP
// using only the Go standard library (net/smtp), no third-party
// dependency needed for this.
package email

import (
	"crypto/tls"
	"fmt"
	"log/slog"
	"net/smtp"
	"strings"

	"github.com/marginpulse/backend/internal/config"
)

// Send delivers a plain-text/HTML email. If SMTP isn't configured
// (SMTP_HOST empty — typical for local dev), it logs the email instead
// of sending, matching the graceful-degradation pattern used elsewhere
// in this app (e.g. insights falling back to a rule-based summary when
// ANTHROPIC_API_KEY is unset) rather than erroring out a request whose
// core purpose (e.g. "create the team invite") already succeeded.
func Send(to, subject, htmlBody string) error {
	cfg := config.Get()
	if cfg.SMTPHost == "" {
		slog.Info("smtp_not_configured_logging_instead", "to", to, "subject", subject)
		return nil
	}

	from := cfg.IngestEmailAddress
	msg := buildMessage(from, to, subject, htmlBody)
	addr := fmt.Sprintf("%s:%d", cfg.SMTPHost, cfg.SMTPPort)

	auth := smtp.PlainAuth("", cfg.SMTPUser, cfg.SMTPPassword, cfg.SMTPHost)

	// Port 465 is implicit TLS (SMTPS); 587 (the common default) uses
	// STARTTLS, which smtp.SendMail negotiates automatically. Both paths
	// are supported since different providers (SendGrid, SES, Gmail)
	// default to different ports.
	if cfg.SMTPPort == 465 {
		return sendImplicitTLS(addr, cfg.SMTPHost, auth, from, to, msg)
	}
	return smtp.SendMail(addr, auth, from, []string{to}, msg)
}

func sendImplicitTLS(addr, host string, auth smtp.Auth, from, to string, msg []byte) error {
	conn, err := tls.Dial("tcp", addr, &tls.Config{ServerName: host})
	if err != nil {
		return err
	}
	defer conn.Close()

	client, err := smtp.NewClient(conn, host)
	if err != nil {
		return err
	}
	defer client.Close()

	if auth != nil {
		if err := client.Auth(auth); err != nil {
			return err
		}
	}
	if err := client.Mail(from); err != nil {
		return err
	}
	if err := client.Rcpt(to); err != nil {
		return err
	}
	w, err := client.Data()
	if err != nil {
		return err
	}
	if _, err := w.Write(msg); err != nil {
		return err
	}
	if err := w.Close(); err != nil {
		return err
	}
	return client.Quit()
}

func buildMessage(from, to, subject, htmlBody string) []byte {
	var b strings.Builder
	b.WriteString("From: " + from + "\r\n")
	b.WriteString("To: " + to + "\r\n")
	b.WriteString("Subject: " + subject + "\r\n")
	b.WriteString("MIME-Version: 1.0\r\n")
	b.WriteString("Content-Type: text/html; charset=UTF-8\r\n")
	b.WriteString("\r\n")
	b.WriteString(htmlBody)
	return []byte(b.String())
}

// SendTeamInvite mirrors email_service.py's send_team_invite_email().
func SendTeamInvite(toEmail, businessName, inviteToken, role string) error {
	subject := fmt.Sprintf("You've been invited to join %s on MarginPulse Pro", businessName)
	body := fmt.Sprintf(
		`<p>You've been invited to join <strong>%s</strong> on MarginPulse Pro as a <strong>%s</strong>.</p>
<p><a href="https://app.marginpulse.io/accept-invite?token=%s">Accept invitation</a></p>`,
		businessName, role, inviteToken,
	)
	return Send(toEmail, subject, body)
}

File: backend-cloudflare/src/email.ts
import type { Env } from "./config";

// DELIBERATE PROVIDER CHANGE, not a straight port: internal/email/email.go
// spoke raw SMTP (net/smtp, including a hand-rolled STARTTLS/implicit-TLS
// path) directly over a TCP socket. Workers can open raw TCP sockets via
// `cloudflare:sockets`, but reimplementing the SMTP protocol by hand over
// that socket (EHLO, STARTTLS upgrade, AUTH PLAIN, MAIL FROM/RCPT TO/DATA)
// is real protocol work that's easy to get subtly wrong and hard to test
// from this environment — not something to fake as "done."
//
// This uses Resend's HTTP API instead (https://resend.com/docs/api-reference/emails/send-email)
// — one fetch() call, no TCP/TLS/SMTP protocol handling needed, which is
// the standard pattern for sending email from Workers. Swap the fetch
// target/body shape here if you'd rather use Postmark, SendGrid, or
// Mailgun's HTTP APIs instead; the graceful "not configured" fallback
// behavior below is preserved exactly as the Go version had it.
export async function sendEmail(env: Env, to: string, subject: string, htmlBody: string): Promise<void> {
  if (!env.RESEND_API_KEY) {
    console.log("email not configured — logging instead of sending", { to, subject });
    return;
  }
  const resp = await fetch("https://api.resend.com/emails", {
    method: "POST",
    headers: { Authorization: `Bearer ${env.RESEND_API_KEY}`, "Content-Type": "application/json" },
    body: JSON.stringify({ from: env.INGEST_EMAIL_ADDRESS, to, subject, html: htmlBody }),
  });
  if (!resp.ok) {
    console.error("email send failed", { status: resp.status, body: await resp.text() });
  }
}

export async function sendTeamInvite(env: Env, toEmail: string, businessName: string, inviteToken: string, role: string): Promise<void> {
  const subject = `You've been invited to join ${businessName} on MarginPulse Pro`;
  const body = `<p>You've been invited to join <strong>${businessName}</strong> on MarginPulse Pro as a <strong>${role}</strong>.</p>
<p><a href="${env.FRONTEND_ORIGIN}/accept-invite?token=${inviteToken}">Accept invitation</a></p>`;
  await sendEmail(env, toEmail, subject, body);
}
